### PR TITLE
Clustergen: updates for clusterclass testing

### LINKED ISSUE #1339: Leverage clustergen-generated configurations to evaluate ClusterClass suitability
**Describe the feature request**

Clustergen testcases, in particular the positive ones (those that generates valid cluster manifests), serves a good dataset of cluster configuration we want to aim to provide equivalent functionality via ClusterClass. 
Ideally we can provide a means (in conjunction with #1337) to automate the evaluation of the CCs against this dataset or some subset of it, so as to monitor improvements or regressions to the CC quality.

**Describe alternatives you've considered**
Create one-off configurations to validate against CC being developed.

**Affected product area (please put an X in all that apply)**

- * CLI
- * Cluster Lifecycle

**Additional context**

### PR DESCRIPTION
**Pull request by vuil**
### What this PR does / why we need it

```
    the 'clustergen' make target will generate outputs for non-clusterclass and
    clusterclass-based configuration

    CI invocation scripts will leverage that to produce
    - the before/after diffs of non-cc-based configs
    - the before/after diffs of cc-based configs
    - the diffs of outputs between the non-cc-based config and its cc-based
    equivalent

    o runtests.sh mods
    - generate cc-based outputs
    - pass and filter by infra
    - remove cconly template patch, but instead,
    - expect processing of a legitimate clusterclass-ccname.yaml

    o for dry-run testing clusterclass-template yamls are included so that the ytt processor 
       can retrieve the cluster class yamls, but they will be removed since the default cluster 
       classes are expected to exist on the management cluster

    o clustergen: update invocation scripts
    to support diffing ClusterClass-based generation against
    non-ClusterClass-based outputs

    o update param_models files

    o generate resources with unrandomized names during dry-run for cleaner comparisons
```

### Which issue(s) this PR fixes

Fixes #1339

### Describe testing done for PR

```make clustergen```
Also will be verifying the diffs in the CI clustergen diffs comment entry

### Release note

```release-note
None
```

### PR Checklist

- x Squash the commits into one or a small number of logical commits
      
- x Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- x Ensure PR contains terms all contributors can understand and links all contributors can access

### CODE REVIEW

File: pkg/v1/providers/infrastructure-azure/v1.0.1/cconly/overlay.yaml [hunk 1]
```diff
@@ -0,0 +1,61 @@
+#@ load("@ytt:overlay", "overlay")
```
Comment (saji-pivotal) [inline pkg/v1/providers/infrastructure-azure/v1.0.1/cconly/overlay.yaml:1]: @vuil Since we've merged CAPZ v1.2.1 to main, this either needs a rebase, or all Azure related changes should go in as a separate PR

File: pkg/v1/tkg/client/cluster.go [hunk 2]
```diff
@@ -61,8 +62,44 @@ type waitForAddonsOptions struct {
...
+	if isCustomOverlayDetected && !isSuppressProvidersUpdateSet {
+		return c.legacyClusterCreation(options, waitForCluster)
+	}
+
+	// TODO: Add steps required to create a CC based cluster
+	return nil
+}
+
+// Create cluster without using ClusterClasses, if custom overlays or modifications to the overlays are detected
+func (c *TkgClient) legacyClusterCreation(options *CreateClusterOptions, waitForCluster bool) error { //nolint:gocyclo,funlen
```
Comment (saji-pivotal) [inline pkg/v1/tkg/client/cluster.go:101]: nit: `s/legacyClusterCreation/legacyClusterCreate`

**Review by saji-pivotal — CHANGES_REQUESTED** [review]

File: pkg/v1/tkg/client/cluster.go [hunk 3]
```diff
@@ -61,8 +62,44 @@ type waitForAddonsOptions struct {
...
+func (c *TkgClient) CreateCluster(options *CreateClusterOptions, waitForCluster bool) error {
+	var isCustomOverlayDetected bool
+	var err error
+
+	if isCustomOverlayDetected, err = c.isCustomOverlayPresent(); err != nil {
+		return err
+	}
+
+	// Ignore the checksum check if config variable 'SUPPRESS_PROVIDERS_UPDATE' is set.
+	isSuppressProvidersUpdateSet := os.Getenv(constants.SuppressProvidersUpdate) != ""
```
[the remark this answers is not in the document]
Comment (vuil) [reply pkg/v1/tkg/client/cluster.go:90]: no it has nothing to do with the ytt processing, so should not be in config_default.yaml. It is a preexisting envvar used during testing to enable testing of local provider overlays changes.

File: pkg/v1/tkg/client/cluster.go (hunk 2 above)
Comment (vuil) [reply pkg/v1/tkg/client/cluster.go:101]: This will be addressed in a subsequent refactoring, so skipping this change for now.

File: pkg/v1/tkg/client/cluster.go [hunk 4]
```diff
@@ -61,8 +62,44 @@ type waitForAddonsOptions struct {
...
+	return providersChecksum != "" && providersChecksum == prePopulatedChecksumFromFile, nil
+}
+
 // CreateCluster create workload cluster
-func (c *TkgClient) CreateCluster(options *CreateClusterOptions, waitForCluster bool) error { //nolint:gocyclo,funlen
+func (c *TkgClient) CreateCluster(options *CreateClusterOptions, waitForCluster bool) error {
+	var isCustomOverlayDetected bool
+	var err error
+
+	if isCustomOverlayDetected, err = c.isCustomOverlayPresent(); err != nil {
```
[the remark this answers is not in the document]
Comment (vuil) [reply pkg/v1/tkg/client/cluster.go:85]: Thanks, the rename will be addressed this in the followup refactor PR

File: pkg/v1/providers/infrastructure-azure/v1.0.1/cconly/overlay.yaml (hunk 1 above)
Comment (vuil) [reply pkg/v1/providers/infrastructure-azure/v1.0.1/cconly/overlay.yaml]: updatec

**Review by saji-pivotal — APPROVED** [review]
There are a few CodeCov complaints that we should probably address. But I would be ok adding tests in a subsequent PR as well.

File: pkg/v1/providers/tests/clustergen/README_ccgen.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+## Introduction
+
+The clustergen tool has been augmented with the ability to generate cluster
+resources using the old (pure client-side ytt-processing) and the new
+(clusterclass-based using dryrun functionality) from the same set of cluster
+configuration file
+
+```text
+make clustergen-cc
```
Comment (tenczar) [inline pkg/v1/providers/tests/clustergen/README_ccgen.md:11]: this is just clustergen now right?

**Review by tenczar — APPROVED** [review]
LGTM

File: pkg/v1/providers/tests/clustergen/README_ccgen.md (hunk 5 above)
Comment (vuil) [reply pkg/v1/providers/tests/clustergen/README_ccgen.md]: good catch, update.

**Review by anujc25 — APPROVED** [review]

**Comment by alfredthenarwhal** [comment]
> @navidshaikh:  /test install-vc7
Commit: 0dff025

Tests failed! Build no: 2137

**Comment by alfredthenarwhal** [comment]
> @rajaskakodkar:  /test install-aws
Commit: 0dff025

Tests failed! Build no: 2143

**Comment by alfredthenarwhal** [comment]
> @navidshaikh:  /test install-vc7
Commit: 0dff025

Tests failed! Build no: 2144

**Comment by alfredthenarwhal** [comment]
> @vuil:  /test install-vc7
Commit: 0dff025

Tests failed! Build no: 2150

**Comment by alfredthenarwhal** [comment]
> @vuil:  /test install-vc7
Commit: 0ae386f

Tests passed! Build no: 2155
